keps/0025-template-to-yaml-function.md

```diff
@@ -0,0 +1,139 @@
+creation-date: 2020-02-25
+last-updated: 2020-02-25
+status: implementable
+see-also:
+  - KEP-24
+---
+
+# Template toYaml function
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+    * [Deep transcription field-by-field](#deep-transcription-field-by-field)
+
+## Summary
+
+This KEP describes addition of a function called `toYaml` for use in the operator templates.
+
+## Motivation
+
+This enhancement will make it possible for operator developers to allow users to
+pass arbitrary structured parameters and directly embed them into resources.
+
+### Goals
+
+- Make it possible for an operator developer to use a `toYaml` function in their templates
+on a `.Params` value.
+
+### Non-Goals
+
+- Make it possible for an operator user to provide a parameter with value type other
+than `string`. We rely on [KEP-24](https://github.com/kudobuilder/kudo/pull/1356) to provide
+this functionality.
+- Make it possible for a user to provide a complex structured parameter value in
+typical multi-line YAML syntax in a convenient way.
+This will be addressed by a separate KEP. For now, a user will need to rely on
+passing the value via a `--parameter` command line flag using JSON syntax
+(which is a subset of YAML).
+
+## Proposal
+
+The idea is to provide an analog of the `toYaml`
+[function provided by Helm](https://github.com/helm/helm/blob/be1e974cccec4f5583ef6e67b229f35f9e6edd2e/pkg/chartutil/files.go#L168-L179).
+Sadly, I could not find any documentation for it, but it is [used frequently in Helm charts](https://github.com/helm/charts/search?q=toYaml&unscoped_q=toYaml).
+
+### User Stories
+
+There were at least a couple of feature requests that would be solved by this
+enhancement. They were mostly concerned with configuring some aspects of k8s pods:
+- [#1221](https://github.com/kudobuilder/kudo/issues/1221): `nodeSelector`, `podAnnotations`, `resources` and `podAffinity`
+- [#951](https://github.com/kudobuilder/kudo/issues/951) `nodeAffinity` and `tolerations`
+
+The common theme is that these k8s features requires deeply nested YAML syntax
+([some examples](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity) are 5 levels deep!)
+
+Here is how using this might look like:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    metadata:
+    {{- if or .Params.podAnnotations .Params.enableMetrics }}
+      annotations:
+      {{- if .Params.enableMetrics }}
+        prometheus.io/scrape: "true"
+      {{- end }}
+      {{- if .Params.podAnnotations }}
+{{ toYaml .Params.podAnnotations | trim | indent 8 }}
```

Comment:
One question here, why do we need the `trim` and the `indent 8`? Is that required for the toYaml to work correctly? And does the `{{ ... }}` block has to be unindented? 

Would be nice if we could figure out a way to do that automatically...

Reply:
Regarding the de-dent and the `| indent 8` - AFAICT this is a limitation of the templating engine. Functions do not have access to the context they are invoked in, so there is no way for `toYaml` to know how much its `{{` is indented in order to replicate this on subsequent lines of its output.

As for the `trim`, here's the [story](https://github.com/helm/helm/issues/3470) with all the context. As the [linked issue](https://github.com/go-yaml/yaml/issues/355#issuecomment-379044221) mentions, this kind of goes hand-in-hand with the `indent`.

Reply:
Added a link to this discussion in the doc itself.